utils_cv/action_recognition/model.py

```diff
@@ -317,19 +371,307 @@ def train_an_epoch(
+                + "</p>"
+            )
+
+            # Plot final results nicely
+            update_println(println)
+            scores_sum -= scores_cache.popleft()
+
+        # Inference done. Ready to run on the next frames.
+        window.popleft()
+        if is_ready:
+            is_ready[0] = True
+
+    def predict_video(
+        self,
+        video_fpath: str,
+        labels: List[str] = None,
+        averaging_size: int = 5,
+        score_threshold: float = 0.025,
+        target_labels: List[str] = None,
+        transforms: Compose = None,
+    ) -> None:
+        """Load video and show frames and inference results while displaying the results
+        """
+        # set model device and to eval mode
+        self.model.to(torch_device())
+        self.model.eval()
+
+        # set up video reader
+        video_reader = decord.VideoReader(video_fpath)
+        print(f"Total frames = {len(video_reader)}")
+
+        # set up ipython jupyter display
+        d_video = IPython.display.display("", display_id=1)
+        d_caption = IPython.display.display("Preparing...", display_id=2)
+
+        # set vars
+        is_ready = [True]
+        window = deque()
+        scores_cache = deque()
+
+        # use labels if given, else see if we have labels from our dataset
+        if not labels:
+            if self.dataset.classes:
+                labels = self.dataset.classes
+            else:
+                raise ("No labels found, add labels argument.")
+        scores_sum = np.zeros(len(labels))
+
+        # set up transforms
+        if not transforms:
+            transforms = get_transforms(train=False)
+
+        # set up print function
+        def update_println(println):
+            d_caption.update(IPython.display.HTML(println))
+
+        while True:
+            try:
+                frame = video_reader.next().asnumpy()
+                if len(frame.shape) != 3:
+                    break
+
+                # Start an inference thread when ready
+                if is_ready[0]:
+                    window.append(frame)
+                    if len(window) == self.sample_length:
+                        is_ready[0] = False
+                        Thread(
+                            target=self.predict_frames,
+                            args=(
+                                window,
+                                scores_cache,
+                                scores_sum,
+                                is_ready,
+                                averaging_size,
+                                score_threshold,
+                                labels,
+                                target_labels,
+                                transforms,
+                                update_println,
+                            ),
+                        ).start()
+
+                # Show video preview
+                f = io.BytesIO()
+                im = Image.fromarray(frame)
```

Comment:
[minor] Fyi, the widget is flickering heavily on my machine. Simply resizing the frame here before calling d_video.update() seems to help significantly with that.
               w,h = frame.shape[0], frame.shape[1]
                scale = 300.0/max(w,h)
                w = round(w*scale)
                h = round(h*scale)
                im = im.resize((h,w))

Reply:
done